### PR TITLE
New search algorithms

### PR DESCRIPTION
**Pull request by Yard1**
# New search algorithms

## What

This PR adds support for three new search algorithms included in [`tune.suggest`](https://docs.ray.io/en/master/tune/api_docs/suggestion.html) - [HyperOpt](https://docs.ray.io/en/master/tune/api_docs/suggestion.html#tune-hyperopt) (`"hyperopt"`), [BOHB](https://docs.ray.io/en/master/tune/api_docs/suggestion.html#suggest-tunebohb) (`"bohb"`) and [Optuna](https://docs.ray.io/en/master/tune/api_docs/suggestion.html#tune-optuna) (`"optuna"`). The support for Optuna is temporarily disabled, as it is not yet present in a stable version of `tune`.

All three new search algorithms take the same param grids as `"bayesian"` did, with the exception of `scikit-optimize` objects.

Due to the new choices present, `scikit-optimize` has been removed from hard dependencies. A user friendly message and docstrings explain how to install required libraries for each search type.

Early stopping has been offloaded onto schedulers that support it.

Additionally, this PR fixes an exception that may occur with certain schedulers (like normal HyperBand) - `36445c373d02089e5d9459152023a51161f8a94c` - and makes small changes to `_Trainable` to support new search algorithms - `1f56318ffff0d309333e80191d58ffef299cc194`

## Why

`scikit-optimize` is not yet updated for the newest `scikit-learn` version, and thus cannot be used. A selection of alternative search algorithms not only solves that issue, but gives greater choice to the user. There is demand for such an enhancement, as evidenced by https://github.com/ray-project/tune-sklearn/issues/65 - and as a contributor to [PyCaret](https://github.com/pycaret/pycaret/) library, we would like to use `tune-sklearn` for Bayesian optimization, but cannot due to `scikit-optimize` being outdated. 

## How

I will add annotations to lines of code that may require them.

## Notes

Code has been formatted with the provided utility. I have not had the chance to add examples or tests. If the maintainers consider this PR a good idea, I will happily add them, as to not waste time right now. I have tested the newly added algorithms locally and they work as expected.

This could also benefit from adding support for conditional search spaces - if this PR is accepted, I will add them, too.

### CODE REVIEW

**Comment by Yard1** [comment]
Ok, I give up on the tests. Everything works fine, aside from scikit-optimize erroring out, which was the impulse to make this

**Comment by richardliaw** [comment]
@Yard1 Awesome, thanks for the PR! Let me take a pass.

File: tune_sklearn/_trainable.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ def _setup(self, config):
         self.return_train_score = config.pop("return_train_score")
         self.n_jobs = config.pop("n_jobs")
         self.estimator_config = config
+        self.pickled = False
```
Comment (richardliaw) [inline tune_sklearn/_trainable.py:50]: question: do we need this?

File: tune_sklearn/_trainable.py [hunk 2]
```diff
@@ -196,4 +197,5 @@ def _restore(self, checkpoint):
             warnings.warn("No estimator restored")
 
     def reset_config(self, new_config):
-        return False
+        self.config = new_config
+        return True
```
Comment (richardliaw) [inline tune_sklearn/_trainable.py:201]: I think this won't work in general, since some support early stopping.

File: tune_sklearn/tune_basesearch.py [hunk 3]
```diff
@@ -225,16 +225,16 @@ def __init__(self,
...
                         self.early_stopping = HyperBandScheduler(
-                            metric="average_test_score")
+                            metric="average_test_score", max_t=max_iters)
                     elif early_stopping == "MedianStoppingRule":
                         self.early_stopping = MedianStoppingRule(
                             metric="average_test_score")
                     elif early_stopping == "ASHAScheduler":
                         self.early_stopping = ASHAScheduler(
-                            metric="average_test_score")
+                            metric="average_test_score", max_t=max_iters)
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:237]: looks good!

File: tune_sklearn/tune_basesearch.py [hunk 4]
```diff
@@ -492,7 +492,7 @@ def _format_results(self, n_splits, out):
 
         """
         dfs = list(out.fetch_trial_dataframes().values())
-        finished = [df[df["done"]] for df in dfs]
+        finished = [df.iloc[[-1]] for df in dfs]
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:495]: what's the reason for this change?

File: tune_sklearn/tune_search.py [hunk 5]
```diff
@@ -349,4 +543,70 @@ def _tune_run(self, config, resources_per_trial):
...
+                scheduler=self.early_stopping,
+                reuse_actors=True,
+                verbose=self.verbose,
+                stop=stop_condition,
+                num_samples=self.num_samples,
+                config=config,
+                fail_fast=True,
+                checkpoint_at_end=True,
+                resources_per_trial=resources_per_trial,
+                local_dir=os.path.expanduser(self.local_dir))
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:610]: can we reuse this execution logic with the switch-cases above?

File: tune_sklearn/tune_search.py [hunk 6]
```diff
@@ -293,10 +474,18 @@ def _tune_run(self, config, resources_per_trial):
...
             config["estimator"] = [
                 clone(self.estimator) for _ in range(self.n_splits)
             ]
+            if hasattr(self.early_stopping, '_max_t_attr'):
+                # we want to delegate stopping to schedulers which support it, but we want it to stop eventually, just in case
+                # the solution is to make the stop condition very big
+                stop_condition = {
+                    "training_iteration": self.max_iters * self.num_samples *
+                    1000
+                }
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:488]: Can you leave this as a TODO? `_max_t_attr` is not public API, but I think it's safer to leave as-is for now.

File: tune_sklearn/tune_search.py [hunk 7]
```diff
@@ -176,18 +192,68 @@ def __init__(self,
...
+        if search_optimization == "bayesian":
+            try:
+                import skopt
+                from skopt import Optimizer
+                from ray.tune.suggest.skopt import SkOptSearch
+            except:
+                traceback.print_exc(file=sys.stderr)
+                print(
+                    "It appears that scikit-optimize is not installed. Do: pip install scikit-optimize",
+                    file=sys.stderr)
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:223]: can we do `logger.execption(); logger.error(msg)`?

File: tune_sklearn/tune_search.py [hunk 8]
```diff
@@ -176,18 +192,68 @@ def __init__(self,
...
+        elif search_optimization == "optuna":
+            try:
+                from ray.tune.suggest.optuna import OptunaSearch, param
+                import optuna
+            except:
+                traceback.print_exc(file=sys.stderr)
+                print(
+                    "It appears that optuna is not installed. Do: pip install optuna",
+                    file=sys.stderr)
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:251]: nit: can we move this library-checking validation logic into a `_try_import_library()`? same with the above

**Comment by richardliaw** [comment]
btw this looks great! Can you add some tests for the new search algorithms?

**Comment by Yard1** [comment]
Will add some tests later, sure.

**Comment by rohan-gt** [comment]
@Yard1 Not sure if it helps in this workflow but I just stumbled into the Sklearn wrapper for Optuna: [OptunaSearchCV](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.OptunaSearchCV.html#optuna-integration-optunasearchcv). It seems to be experimental but is more robust (doesn't error out) compared to scikit-optimize based on the tests that I performed

**Comment by Yard1** [comment]
@rohan-gt Oh, that's great. It's not really applicable here, but we may use it for PyCaret. The question of how to exactly implement it is still in the air.

**Comment by Yard1** [comment]
Added examples using HpBandSter and Hyperopt, as well as allowed the passing of parameters in algorithm-specific formats for greater control.

File: tune_sklearn/_trainable.py (hunk 1 above)
Comment (richardliaw) [reply tune_sklearn/_trainable.py:50]: nvm, checked code. Thanks!

File: tune_sklearn/tune_basesearch.py (hunk 4 above)
Comment (richardliaw) [reply tune_sklearn/tune_basesearch.py:495]: nvm, checked code. Thanks!

**Comment by Yard1** [comment]
@richardliaw Would you like me to update the readme, too?

**Comment by richardliaw** [comment]
That'd be great :) Thanks a bunch!

**Comment by richardliaw** [comment]
@Yard1 I will merge this in ~3 hours; if we don't get the README updates in by then, feel free to open another PR!

**Comment by Yard1** [comment]
@richardliaw Updated the readme, please look it over. I may have gone a bit too in-depth.

File: README.md [hunk 9]
```diff
@@ -20,9 +20,12 @@ If the estimator does not support `partial_fit`, a warning will be shown saying
 `pip install tune-sklearn ray[tune]`
 
 ## Examples
+
 #### TuneGridSearchCV
 To start out, it’s as easy as changing our import statement to get Tune’s grid search cross validation interface, and the rest is almost identical!
 
+`TuneGridSearchCV` accepts dictionaries in the format `{ param_name:str : distribution: list }` or a list of such dictionaries, just like scikit-learn's `GridSearchCV`.
```
Comment (richardliaw) [inline README.md:27]:
```suggestion
`TuneGridSearchCV` accepts dictionaries in the format `{ param_name: str : distribution: list }` or a list of such dictionaries, just like scikit-learn's `GridSearchCV`.
```

**Comment by richardliaw** [comment]
Looks fine to me! Thanks a bunch @Yard1!
